### DISCUSSION TITLE
not working

### DISCUSSION THREAD

**Discussion by abdullrahmanesmael07** [jomjol/AI-on-the-edge-device#3039] [General]
for some reasion its not working with me it says failed and some libarys dont exist like psram

**Comment by SybexX** [comment]
What is not working, compiling the firmware????????????
If you want to get help, you have to give more details here.

**Reply by abdullrahmanesmael07** [reply]
na for some reason i only get FAILED, i tried to ask my teacher, but he couldnt solve it either

**Reply by SybexX** [reply]
Take a picture of it and post it here.

**Reply by SybexX** [reply]
I couldn't tell from your statements whether you want to compile your own firmware or whether you're trying to flash the ESP.
when it comes to compiling:
The libraries code/components/esp32-camera, code/components/esp-nn, code/components/esp-tflite-micro and code/components/stb 
are not downloaded automatically and must be downloaded separately from the respective pages.

![Unbenannt](https://github.com/jomjol/AI-on-the-edge-device/assets/587201/c1a28dd1-5ad5-450a-8289-d77cbeaa87e5)

All the libraries you need are included here 
[AI-on-the-edge-device-rolling.zip](https://github.com/jomjol/AI-on-the-edge-device/files/15084722/AI-on-the-edge-device-rolling.zip)

**Comment by caco3** [comment]
Just use the precompiled files as @SybexX already stated.

If you want to compile it yourself, please have a look and **exactly** follow the instructions in https://github.com/jomjol/AI-on-the-edge-device/tree/rolling/code#readme
How ever, the project is not trivial, if you want to learn coding, this project is IMO not a good starting point!

**Reply by Abdullrahman4IT** [reply]
i alr managed to solve it thx.

although my teacher wants me to read the values and make a website on my own where i can connect the esp to any wifi and read the values of the website. any tips

**Comment by SybexX** [comment]
Do I understand correctly that you want to have a website on the ESP where you can select the WLAN and enter the associated access data?

you could use the following to read values:
IP of your ESP/info
IP of your ESP/sysinfo
IP of your ESP/starttime
IP of your ESP/statusflow
IP of your ESP/cpu_temperature
IP of your ESP/rssi
IP of your ESP/uptime
IP of your ESP/value
IP of your ESP/json
IP of your ESP/heap

e.g. so: 192.168.178.100/info

**Reply by abdullrahmanesmael07** [reply]
let me clarify.

we need a website created in either Vue or just plain css, html and js.

### Use Cases and User Stories

- As a user, I want to control the device via web application on a smart phone as well as on a desktop PC.
- As a user, I want to configure the device for a specific gauge via web application.
- As a user, I want to train the device for a specific gauge via web application.
- As a user, I want to read the recognized value via web application. If feasible, I want to view the images from which the value was recognized.
- As a user, I want to integrate the device into my WiFi network easily.
- As a user, I want that the device time is synchronized with the real time of my timezone.
- As a user, I want to monitor the device via MQTT.
- As a user, I want to update the device software over the air.


### Tasks

- Develop a software design showing the major software components and their relations as well as required communication protocols.
- Integrate a WiFi setup and configuration on the device
- Integrate a web server on the device.
- Integrate an SNTP client on the device, including SNTP configuration via integrated web server.
- Integrate the AI image recognition software on the device.
- Extend the AI image recognition as required by user stories.
- Implement a web service on the device for manually reading the last recognized values.
- Implement a web service on the device for configuring the device.
- Implement a web service on the device for train the AI on the device.
- Implement the required front-end.
- Implement an MQTT-based monitoring of the device status.
- Implement an MQTT-based monitoring the recognized value.
- Integrate and enable SWUOTA (software update over the air).

there is already this https://jomjol.github.io/AI-on-the-edge-device/index.html, so i assume i could maybe use webscraping to get the data, but im not sure

keep in mind, that I am only an beginner in programming.


thanks for everyone that helps

**Reply by caco3** [reply]
Hi @abdullrahmanesmael07

I am quite surprised!
What kind of school is this? And how skilled is your teacher?
The user stories you have are quite complex and cover a lot of things. This clearly is not in the scope of a normal diploma thesis!
For me, who has quite some experience, it would take me months to do such thing! And you say, you are a beginner in programming. SO it would even take you much longer.

We showed you already that there is a REST API, you can create a website on a different device and fetch the information from there, if you need. But actively controlling and configuring the device from there, that's for sure out of scope for a beginner!

> Implement a web service on the device for train the AI on the device.

Note that this is not possible at all on such device. The training needs to be on a more powerful system. The ESP then only uses the generated model.

**Reply by abdullrahmanesmael07** [reply]
well, thats what he wants from us. he himself says he doesnt know, so i guess i will fail this class.

**Reply by caco3** [reply]
> well, thats what he wants from us. he himself says he doesnt know, so i guess i will fail this class.

A good teacher is able to solve the exercises he gives to his students without issues.
If he really thinks you should be able to solve this exercise by yourself, just point him to this discussion and we can explain it to him once more.

**Reply by abdullrahmanesmael07** [reply]
how can i use the REST API

**Reply by caco3** [reply]
That, I think, you should be able to find out by yourself.
@SybexX already showed you a list of endpoints, see above.
Also see https://jomjol.github.io/AI-on-the-edge-device-docs/REST-API/

**Reply by abdullrahmanesmael07** [reply]
ok thanks, another question. how can i see the files of the ESP? and how can i change the sytle of the html. because i noticed that the site has all the requirements of my teacher. so if i can change the style.css it would be better

**Reply by caco3** [reply]
All files of the Web interface are placed on the sd card in the folder `html` you can edit them as you want.
The files on the sd card can be deleted and uploaded with the file manager in the menu `System > File Server`.

For easier playing with the web interface, you can also run it locally on your PC. Just copy all files from the folder `html` on the release file (https://github.com/jomjol/AI-on-the-edge-device/releases/download/v15.7.0/AI-on-the-edge-device__update__v15.7.0.zip) to yout PC and edit the top of `common.js`:
```js
/* The UI can also be run locally, but you have to set the IP of your devide accordingly.
 * And you also might have to disable CORS in your webbrowser!
 * Keep empty to disable using it. Enabling it will break access through a forwared port, see 
 * https://github.com/jomjol/AI-on-the-edge-device/issues/2681 */
var domainname_for_testing = "";
//var domainname_for_testing = "192.168.1.151";
```

Then open the `index.html` in your webbrowser.
Also read the notes regarding `CORS`!

**Reply by abdullrahmanesmael07** [reply]
i thank you sooo much caco3!! I am kinda ashamed that i am asking so much. how did you find about that? like did you work with jomjol on the project or what?

**Reply by caco3** [reply]
> did you work with jomjol on the project or what?

I am one of the developers.

**Reply by abdullrahmanesmael07** [reply]
Good morning gentleman, i have a question! Is there any possible way to edit the wifi.ini, while the ESP is running or is it not possible.

Best regards and thank you

**Reply by SybexX** [reply]
this is prevented in the "AI-on-the-edge-device-rolling.../code/components/jomjol_fileserver_ota/server_file.cpp".
With the following customized file, you should be able to access the wlan.ini via the file server: [jomjol_fileserver_ota.zip](https://github.com/user-attachments/files/15882937/jomjol_fileserver_ota.zip)
copy the files in the zip to AI-on-the-edge-device-rolling.../code/components/jomjol_fileserver_ota/ and compile again
![wlan_ini](https://github.com/jomjol/AI-on-the-edge-device/assets/587201/183b1f59-7965-42e5-947b-607d57b76116)

**Reply by abdullrahmanesmael07** [reply]
did i understand correctly, i just have to download the zip then go to file server and upload this zip and i am good to go?
note that i am using the online flasher.

**Reply by caco3** [reply]
no, you have to use the zip file to do the Over-The-Air (OTA) update.
see the menu 'System'

**Reply by SybexX** [reply]
@abdullrahmanesmael07 As far as I understand, you want to create your own wifi.html in which you can easily and quickly change the wifi access data, is that correct or is my above approach enough for you?

**Reply by abdullrahmanesmael07** [reply]
pretty much yea

**Reply by abdullrahmanesmael07** [reply]
i think you misunderstood me, i do need to make my own wifi.html

**Reply by SybexX** [reply]
@abdullrahmanesmael07 A little help/Example, but I can't say whether everything works as it should because I don't have much time at the moment.

for update via OTA:
[AI-on-the-edge-device__update__rolling_(cb3eb3b).zip](https://github.com/user-attachments/files/15954158/AI-on-the-edge-device__update__rolling_.cb3eb3b.zip)

Project source code:
[AI-on-the-edge-device-edit_wlan_config.zip](https://github.com/user-attachments/files/15954162/AI-on-the-edge-device-edit_wlan_config.zip)

![wifi](https://github.com/jomjol/AI-on-the-edge-device/assets/587201/48b17cb1-ef0c-47f0-850b-f0c01316050b)

**Reply by abdullrahmanesmael07** [reply]
hey thank you, please there is no need to help me if you dont have time. i dont mind it, but i have another question in general. how can i delete whats on the ESP, because i just used the website to download AI on the egde. but i need to flash it via platformio

**Reply by SybexX** [reply]
![Unbenannt1](https://github.com/jomjol/AI-on-the-edge-device/assets/587201/61dd58d9-1150-45f9-891c-7ac1fe0def8c)
build the project:
![Unbenannt2](https://github.com/jomjol/AI-on-the-edge-device/assets/587201/1d11a429-84c4-4a17-b4f9-6a867c01faee)
upload the firmware to the ESP32:
![Unbenannt3](https://github.com/jomjol/AI-on-the-edge-device/assets/587201/3e7701ef-826e-48a7-84f8-f4f1d7a6ae72)
erase the flash from the ESP32:
![Unbenannt4](https://github.com/jomjol/AI-on-the-edge-device/assets/587201/c67216e0-167c-40d1-971f-2ff529e3fbe6)

**Reply by abdullrahmanesmael07** [reply]
thank you so so so much

**Reply by abdullrahmanesmael07** [reply]
thank you so so so much

**Reply by SybexX** [reply]
@abdullrahmanesmael07 I made a few improvements and fixed some errors

for update via OTA:
[AI-on-the-edge-device__update__rolling_(cb3eb3b).zip](https://github.com/user-attachments/files/15954179/AI-on-the-edge-device__update__rolling_.cb3eb3b.zip)

Project source code:
[AI-on-the-edge-device-edit_wlan_config.zip](https://github.com/user-attachments/files/15954180/AI-on-the-edge-device-edit_wlan_config.zip)

![wifi](https://github.com/jomjol/AI-on-the-edge-device/assets/587201/f7bb937c-d5ac-4253-baca-d5ad705406cf)

**Reply by abdullrahmanesmael07** [reply]
thank you so much. you and caco3 literally saved my project

**Reply by abdullrahmanesmael07** [reply]
ok one last question, you know how there is this website https://jomjol.github.io/AI-on-the-edge-device/index.html i use it to connect my esp to the wifi and i can read the ip from the console. but i dont want that. i want to create a website (seperate from the other stuff) that it literally the same as https://jomjol.github.io/AI-on-the-edge-device/index.html but with no console and text fielt, where the ip will be found and a button which takes me to this website https://192.168.164.84.com

i just need help on how to get the ip and that stuff

thank you

**Reply by abdullrahmanesmael07** [reply]
also sorry if it sounds silly, but when you say compiling do you mean flashing the esp?

**Reply by SybexX** [reply]
@abdullrahmanesmael07 compiling = build the project

**Reply by SybexX** [reply]
you can also access it like this:  http:// "the Hostname from ESP32" ------->  http://watermeter

[test.zip](https://github.com/user-attachments/files/15945770/test.zip)

**Reply by abdullrahmanesmael07** [reply]
i mean, it doesnt really work, but thanks anyway. I have finished my project and i will present it tomorrow. Thanks everyone that helped

**Reply by SybexX** [reply]
@abdullrahmanesmael07 You can only access the ESP32 if you know its IP or Hostname.
Unfortunately, there is no other way with html, you had to create a Windows application for it.
Or did you want to be able to use edit_wlan_config.html externally? [test2.zip](https://github.com/user-attachments/files/15952847/test2.zip)
Please note that this only works with my changed firmware, otherwise you will not be able to access the wlan.ini and will only get an error message in the webbrowser console.

**Reply by abdullrahmanesmael07** [reply]
now it doesnt matter, anyway thank you for your help

**Comment by abdullrahmanesmael07** [comment]
Thanks to everyone who helped me finish this project. I am thankful for you help. I hope you all the best.
